cgo: CGO_FLAGS not respected

Compilation of sqlite-3 fails because gb is not passing down the `-ldl` flag
```
% gb build -v -f all                                                                                                                                                                                                                                                                                     
DEBUG project root "/home/dfc/devel/demo4"
DEBUG matchPackages: [/home/dfc/devel/demo4/src] all
DEBUG args: [github.com/mattn/go-sqlite3 github.com/mattn/go-sqlite3/sqlite3_test]
DEBUG loadPackage: github.com/mattn/go-sqlite3 true (/home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3)
DEBUG loadPackage: github.com/mattn/go-sqlite3/sqlite3_test true (/home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3/sqlite3_test)
DEBUG buildPackage: github.com/mattn/go-sqlite3
DEBUG cd /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3; [/home/dfc/go/pkg/tool/linux_amd64/cgo -objdir /tmp/gb251703138/github.com/mattn -importpath github.com/mattn/go-sqlite3 -- -I /tmp/gb251703138/github.com/mattn -I /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3 backup.go error.go sqlite3.go sqlite3_other.go]
DEBUG cd /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3; [gcc -fPIC -m64 -pthread -fmessage-length=0 -I /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3 -I /tmp/gb251703138/github.com/mattn -g -O2 -o /tmp/gb251703138/github.com/mattn/_cgo_main.o -c /tmp/gb251703138/github.com/mattn/_cgo_main.c]
DEBUG cd /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3; [gcc -fPIC -m64 -pthread -fmessage-length=0 -I /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3 -I /tmp/gb251703138/github.com/mattn -g -O2 -o /tmp/gb251703138/github.com/mattn/_cgo_export.o -c /tmp/gb251703138/github.com/mattn/_cgo_export.c]
DEBUG cd /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3; [gcc -fPIC -m64 -pthread -fmessage-length=0 -I /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3 -I . -g -O2 -o sqlite3-binding.o -c sqlite3-binding.c]
DEBUG cd /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3; [gcc -fPIC -m64 -pthread -fmessage-length=0 -I /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3 -I /tmp/gb251703138/github.com/mattn -g -O2 -o /tmp/gb251703138/github.com/mattn/backup.cgo2.o -c /tmp/gb251703138/github.com/mattn/backup.cgo2.c]
DEBUG cd /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3; [gcc -fPIC -m64 -pthread -fmessage-length=0 -I /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3 -I /tmp/gb251703138/github.com/mattn -g -O2 -o /tmp/gb251703138/github.com/mattn/error.cgo2.o -c /tmp/gb251703138/github.com/mattn/error.cgo2.c]
DEBUG cd /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3; [gcc -fPIC -m64 -pthread -fmessage-length=0 -I /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3 -I /tmp/gb251703138/github.com/mattn -g -O2 -o /tmp/gb251703138/github.com/mattn/sqlite3.cgo2.o -c /tmp/gb251703138/github.com/mattn/sqlite3.cgo2.c]
DEBUG cd /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3; [gcc -fPIC -m64 -pthread -fmessage-length=0 -I /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3 -I /tmp/gb251703138/github.com/mattn -g -O2 -o /tmp/gb251703138/github.com/mattn/sqlite3_other.cgo2.o -c /tmp/gb251703138/github.com/mattn/sqlite3_other.cgo2.c]
DEBUG cd /home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3; [gcc -fPIC -m64 -pthread -fmessage-length=0 -o /tmp/gb251703138/github.com/mattn/_cgo_.o /tmp/gb251703138/github.com/mattn/_cgo_main.o /tmp/gb251703138/github.com/mattn/_cgo_export.o sqlite3-binding.o /tmp/gb251703138/github.com/mattn/backup.cgo2.o /tmp/gb251703138/github.com/mattn/error.cgo2.o /tmp/gb251703138/github.com/mattn/sqlite3.cgo2.o /tmp/gb251703138/github.com/mattn/sqlite3_other.cgo2.o -g -O2]
sqlite3-binding.o: In function `unixDlError':
/home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3/sqlite3-binding.c:29939: undefined reference to `dlerror'
sqlite3-binding.o: In function `unixDlClose':
/home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3/sqlite3-binding.c:29970: undefined reference to `dlclose'
sqlite3-binding.o: In function `unixDlSym':
/home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3/sqlite3-binding.c:29966: undefined reference to `dlsym'
sqlite3-binding.o: In function `unixDlOpen':
/home/dfc/devel/demo4/src/github.com/mattn/go-sqlite3/sqlite3-binding.c:29925: undefined reference to `dlopen'
collect2: error: ld returned 1 exit status
# gcc -fPIC -m64 -pthread -fmessage-length=0 -o /tmp/gb251703138/github.com/mattn/_cgo_.o /tmp/gb251703138/github.com/mattn/_cgo_main.o /tmp/gb251703138/github.com/mattn/_cgo_export.o sqlite3-binding.o /tmp/gb251703138/github.com/mattn/backup.cgo2.o /tmp/gb251703138/github.com/mattn/error.cgo2.o /tmp/gb251703138/github.com/mattn/sqlite3.cgo2.o /tmp/gb251703138/github.com/mattn/sqlite3_other.cgo2.o -g -O2
DEBUG buildPackage: github.com/mattn/go-sqlite3/sqlite3_test
DEBUG build duration: 18.276228684s map[]
FATAL command "build" failed: exit status 1
```

--- a/cgo.go
+++ b/cgo.go
@@ -24,7 +24,7 @@ func rungcc1(ctx *Context, dir, ofile, cfile string) error {
 		"-o", ofile,
 		"-c", cfile,
 	}
-	return ctx.run(dir, gcc(), args...)
+	return ctx.run(dir, nil, gcc(), args...)
 }
 
 // rungcc2 links the o files from rungcc1 into a single _cgo_.o.
@@ -35,7 +35,7 @@ func rungcc2(ctx *Context, dir string, ofile string, ofiles []string) error {
 	}
 	args = append(args, ofiles...)
 	args = append(args, "-g", "-O2") // this has to go at the end, because reasons!
-	return ctx.run(dir, gcc(), args...)
+	return ctx.run(dir, nil, gcc(), args...)
 }
 
 // rungcc3 links all previous ofiles together with libgcc into a single _all.o.
@@ -52,7 +52,7 @@ func rungcc3(ctx *Context, dir string, ofiles []string) (string, error) {
 	}
 	args = append(args, ofiles...)
 	args = append(args, "-Wl,-r", "-nostdlib", libgcc, "-Wl,--build-id=none")
-	return ofile, ctx.run(dir, gcc(), args...)
+	return ofile, ctx.run(dir, nil, gcc(), args...)
 }
 
 // libgcc returns the value of gcc -print-libgcc-file-name.
@@ -62,7 +62,7 @@ func libgcc(ctx *Context) (string, error) {
 		"-print-libgcc-file-name",
 	}
 	var buf bytes.Buffer
-	err := ctx.runOut(&buf, ".", gcc(), args...)
+	err := ctx.runOut(&buf, ".", nil, gcc(), args...)
 	return strings.TrimSpace(buf.String()), err
 }
 

--- a/cgo.go
+++ b/cgo.go
@@ -28,14 +28,15 @@ func rungcc1(ctx *Context, dir, ofile, cfile string) error {
 }
 
 // rungcc2 links the o files from rungcc1 into a single _cgo_.o.
-func rungcc2(ctx *Context, dir string, ofile string, ofiles []string) error {
+func rungcc2(pkg *Package, dir string, ofile string, ofiles []string) error {
 	args := []string{
 		"-fPIC", "-m64", "-pthread", "-fmessage-length=0",
 		"-o", ofile,
 	}
 	args = append(args, ofiles...)
-	args = append(args, "-g", "-O2") // this has to go at the end, because reasons!
-	return ctx.run(dir, nil, gcc(), args...)
+	_, _, _, cgoLDFLAGS := cflags(pkg, true)
+	args = append(args, cgoLDFLAGS...) // this has to go at the end, because reasons!
+	return pkg.run(dir, nil, gcc(), args...)
 }
 
 // rungcc3 links all previous ofiles together with libgcc into a single _all.o.

--- a/cgo14.go
+++ b/cgo14.go
@@ -83,7 +83,7 @@ func runcgo1(pkg *Package) error {
 		"-I", pkg.Dir,
 	}
 	args = append(args, pkg.CgoFiles...)
-	return pkg.run(pkg.Dir, cgo, args...)
+	return pkg.run(pkg.Dir, nil, cgo, args...)
 }
 
 // runcgo2 invokes the cgo tool to create _cgo_import.go
@@ -97,5 +97,5 @@ func runcgo2(pkg *Package, ofile string) (string, error) {
 		"-dynimport", ofile,
 		"-dynout", dynout,
 	}
-	return dynout, pkg.run(pkg.Dir, cgo, args...)
+	return dynout, pkg.run(pkg.Dir, nil, cgo, args...)
 }

--- a/cgo14.go
+++ b/cgo14.go
@@ -48,7 +48,7 @@ func cgo(pkg *Package) ([]ObjTarget, []string) {
 	}
 
 	ofile := filepath.Join(filepath.Dir(ofiles[0]), "_cgo_.o")
-	if err := rungcc2(pkg.Context, pkg.Dir, ofile, ofiles); err != nil {
+	if err := rungcc2(pkg, pkg.Dir, ofile, ofiles); err != nil {
 		return fn(ErrTarget{err})
 	}
 

--- a/cgo15.go
+++ b/cgo15.go
@@ -2,7 +2,11 @@
 
 package gb
 
-import "path/filepath"
+import (
+	"path/filepath"
+	"strconv"
+	"strings"
+)
 
 // cgo support functions
 
@@ -41,7 +45,7 @@ func cgo(pkg *Package) ([]ObjTarget, []string) {
 	}
 
 	ofile := filepath.Join(filepath.Dir(ofiles[0]), "_cgo_.o")
-	if err := rungcc2(pkg.Context, pkg.Dir, ofile, ofiles); err != nil {
+	if err := rungcc2(pkg, pkg.Dir, ofile, ofiles); err != nil {
 		return fn(ErrTarget{err})
 	}
 
@@ -75,7 +79,20 @@ func runcgo1(pkg *Package) error {
 		"-I", pkg.Dir,
 	}
 	args = append(args, pkg.CgoFiles...)
-	return pkg.run(pkg.Dir, nil, cgo, args...)
+
+	_, _, _, cgoLDFLAGS := cflags(pkg, true)
+
+	// Update $CGO_LDFLAGS with p.CgoLDFLAGS.
+	var cgoenv []string
+	if len(cgoLDFLAGS) > 0 {
+		flags := make([]string, len(cgoLDFLAGS))
+		for i, f := range cgoLDFLAGS {
+			flags[i] = strconv.Quote(f)
+		}
+		cgoenv = []string{"CGO_LDFLAGS=" + strings.Join(flags, " ")}
+	}
+
+	return pkg.run(pkg.Dir, cgoenv, cgo, args...)
 }
 
 // runcgo2 invokes the cgo tool to create _cgo_import.go

--- a/cgo15.go
+++ b/cgo15.go
@@ -75,7 +75,7 @@ func runcgo1(pkg *Package) error {
 		"-I", pkg.Dir,
 	}
 	args = append(args, pkg.CgoFiles...)
-	return pkg.run(pkg.Dir, cgo, args...)
+	return pkg.run(pkg.Dir, nil, cgo, args...)
 }
 
 // runcgo2 invokes the cgo tool to create _cgo_import.go
@@ -90,5 +90,5 @@ func runcgo2(pkg *Package, ofile string) (string, error) {
 		"-dynimport", ofile,
 		"-dynout", dynout,
 	}
-	return dynout, pkg.run(pkg.Dir, cgo, args...)
+	return dynout, pkg.run(pkg.Dir, nil, cgo, args...)
 }

--- a/context.go
+++ b/context.go
@@ -200,20 +200,21 @@ func (c *Context) Run(cmd *exec.Cmd, dep Target) Target {
 	return &target // TODO
 }
 
-func (c *Context) run(dir, command string, args ...string) error {
+func (c *Context) run(dir string, env []string, command string, args ...string) error {
 	var buf bytes.Buffer
-	err := c.runOut(&buf, dir, command, args...)
+	err := c.runOut(&buf, dir, nil, command, args...)
 	if err != nil {
 		return fmt.Errorf("# %s %s: %v\n%s", command, strings.Join(args, " "), err, buf.String())
 	}
 	return nil
 }
 
-func (c *Context) runOut(output io.Writer, dir, command string, args ...string) error {
+func (c *Context) runOut(output io.Writer, dir string, env []string, command string, args ...string) error {
 	cmd := exec.Command(command, args...)
 	cmd.Dir = dir
 	cmd.Stdout = output
 	cmd.Stderr = os.Stderr
+	cmd.Env = mergeEnvLists(env, envForDir(cmd.Dir))
 	<-c.permits
 	Debugf("cd %s; %s", cmd.Dir, cmd.Args)
 	c.permits <- true
@@ -351,4 +352,33 @@ func matchPackages(c *Context, pattern string) []string {
 		})
 	}
 	return pkgs
+}
+
+// envForDir returns a copy of the environment
+// suitable for running in the given directory.
+// The environment is the current process's environment
+// but with an updated $PWD, so that an os.Getwd in the
+// child will be faster.
+func envForDir(dir string) []string {
+	env := os.Environ()
+	// Internally we only use rooted paths, so dir is rooted.
+	// Even if dir is not rooted, no harm done.
+	return mergeEnvLists([]string{"PWD=" + dir}, env)
+}
+
+// mergeEnvLists merges the two environment lists such that
+// variables with the same name in "in" replace those in "out".
+func mergeEnvLists(in, out []string) []string {
+NextVar:
+	for _, inkv := range in {
+		k := strings.SplitAfterN(inkv, "=", 2)[0]
+		for i, outkv := range out {
+			if strings.HasPrefix(outkv, k) {
+				out[i] = inkv
+				continue NextVar
+			}
+		}
+		out = append(out, inkv)
+	}
+	return out
 }

--- a/gc.go
+++ b/gc.go
@@ -17,5 +17,5 @@ func (t *gcToolchain) Pack(pkg *Package, afiles ...string) error {
 	args := []string{"r"}
 	args = append(args, afiles...)
 	dir := filepath.Dir(afiles[0])
-	return pkg.run(dir, t.pack, args...)
+	return pkg.run(dir, nil, t.pack, args...)
 }

--- a/gc14.go
+++ b/gc14.go
@@ -60,7 +60,7 @@ func (t *gcToolchain) Gc(pkg *Package, searchpaths []string, importpath, srcdir,
 	if err := mkdir(filepath.Dir(outfile)); err != nil {
 		return fmt.Errorf("gc:gc: %v", err)
 	}
-	return pkg.run(srcdir, t.gc, args...)
+	return pkg.run(srcdir, nil, t.gc, args...)
 }
 
 func (t *gcToolchain) Asm(pkg *Package, srcdir, ofile, sfile string) error {
@@ -70,7 +70,7 @@ func (t *gcToolchain) Asm(pkg *Package, srcdir, ofile, sfile string) error {
 	if err := mkdir(filepath.Dir(ofile)); err != nil {
 		return fmt.Errorf("gc:asm: %v", err)
 	}
-	return pkg.run(srcdir, t.as, args...)
+	return pkg.run(srcdir, nil, t.as, args...)
 }
 
 func (t *gcToolchain) Ld(pkg *Package, searchpaths, ldflags []string, outfile, afile string) error {
@@ -83,7 +83,7 @@ func (t *gcToolchain) Ld(pkg *Package, searchpaths, ldflags []string, outfile, a
 	if err := mkdir(filepath.Dir(outfile)); err != nil {
 		return fmt.Errorf("gc:ld: %v", err)
 	}
-	return pkg.run(".", t.ld, args...)
+	return pkg.run(".", nil, t.ld, args...)
 }
 
 func (t *gcToolchain) Cc(pkg *Package, ofile, cfile string) error {
@@ -97,5 +97,5 @@ func (t *gcToolchain) Cc(pkg *Package, ofile, cfile string) error {
 		"-D", "GOARCH_" + pkg.GOARCH,
 		cfile,
 	}
-	return pkg.run(pkg.Dir, t.cc, args...)
+	return pkg.run(pkg.Dir, nil, t.cc, args...)
 }

--- a/gc15.go
+++ b/gc15.go
@@ -61,7 +61,7 @@ func (t *gcToolchain) Gc(pkg *Package, searchpaths []string, importpath, srcdir,
 	if err := mkdir(filepath.Dir(outfile)); err != nil {
 		return fmt.Errorf("gc:gc: %v", err)
 	}
-	return pkg.run(srcdir, t.gc, args...)
+	return pkg.run(srcdir, nil, t.gc, args...)
 }
 
 func (t *gcToolchain) Asm(pkg *Package, srcdir, ofile, sfile string) error {
@@ -71,7 +71,7 @@ func (t *gcToolchain) Asm(pkg *Package, srcdir, ofile, sfile string) error {
 	if err := mkdir(filepath.Dir(ofile)); err != nil {
 		return fmt.Errorf("gc:asm: %v", err)
 	}
-	return pkg.run(srcdir, t.as, args...)
+	return pkg.run(srcdir, nil, t.as, args...)
 }
 
 func (t *gcToolchain) Ld(pkg *Package, searchpaths, ldflags []string, outfile, afile string) error {
@@ -84,7 +84,7 @@ func (t *gcToolchain) Ld(pkg *Package, searchpaths, ldflags []string, outfile, a
 	if err := mkdir(filepath.Dir(outfile)); err != nil {
 		return fmt.Errorf("gc:ld: %v", err)
 	}
-	return pkg.run(".", t.ld, args...)
+	return pkg.run(".", nil, t.ld, args...)
 }
 
 func (t *gcToolchain) Cc(pkg *Package, ofile, cfile string) error {

